### ISSUE TITLE
[Bug][tptr lowering] Gather backward fails on scatter reduce="add" generated tl.atomic_add

### ISSUE THREAD

**Issue by 111a909** [RuyiAI-Stack/triton-riscv#39]
### Triton python code

```mlir
import importlib.util
import os
import threading
import uuid
from pathlib import Path

import torch

from .scatter import scatter_


def write_atomic(
    path_: str,
    content: str,
    make_dirs: bool = False,
    encoding: str = "utf-8",
) -> None:
    path = Path(path_)
    if make_dirs:
        path.parent.mkdir(parents=True, exist_ok=True)
    tmp_path = (
        path.parent
        / f".{os.getpid()}.{threading.get_ident()}.{uuid.uuid4().hex}.tmp"
    )
    with tmp_path.open("wt", encoding=encoding) as f:
        f.write(content)
    tmp_path.replace(path)


def _code_cache_dir():
    d = Path.home() / ".flaggems" / "code_cache"
    d.mkdir(parents=True, exist_ok=True)
    return d


def restride_dim(inp, dim, target_shape):
    shape = list(inp.shape)
    stride = list(inp.stride())
    shape[dim] = target_shape[dim]
    stride[dim] = 0
    for i in range(len(shape)):
        if shape[i] != target_shape[i]:
            if target_shape[i] == 1:
                shape[i] = 1
                stride[i] = 0
            elif shape[i] == 1:
                shape[i] = target_shape[i]
                stride[i] = 0
    return inp.as_strided(shape, stride)


def _generate_imports(code):
    code += "import torch\n"
    code += "import triton\n"
    code += "import triton.language as tl\n\n\n"
    return code


def _generate_gather_kernel(rank, kernel_name, code):
    code += "@triton.jit\n"
    code += f"def {kernel_name}(\n"
    code += "    inp, index, out,\n"
    for i in range(rank):
        code += f"    inp_shape{i},\n"
    for i in range(rank):
        code += f"    index_shape{i},\n"
    for i in range(rank):
        code += f"    out_shape{i},\n"
    for i in range(rank):
        code += f"    inp_stride{i},\n"
    for i in range(rank):
        code += f"    index_stride{i},\n"
    for i in range(rank):
        code += f"    out_stride{i},\n"
    code += "    dim, dim_stride, N,\n"
    code += "    BLOCK_SIZE_N: tl.constexpr,\n"
    code += "):\n"
    code += "    pid = tl.program_id(0)\n"
    code += "    offset = pid * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N).to(tl.int64)\n\n"
    code += "    cur_offset = offset\n"
    for i in range(rank - 1, -1, -1):
        code += f"    index_idx{i} = cur_offset % index_shape{i}\n"
        code += f"    cur_offset = cur_offset // index_shape{i}\n"
    code += "\n"
    comp = " + ".join([f"index_idx{i} * index_stride{i}" for i in range(rank)])
    code += f"    index_offset = {comp}\n"
    code += "    mask = offset < N\n"
    code += (
        "    cur_index = tl.load(index + index_offset, mask=mask, other=0)\n\n"
    )
    comp = " + ".join([f"index_idx{i} * inp_stride{i}" for i in range(rank)])
    code += f"    inp_offset = {comp}\n"
    code += "    inp_offset += cur_index * dim_stride\n"
    code += "    cur_inp = tl.load(inp + inp_offset, mask=mask, other=0)\n\n"
    comp_out = " + ".join(
        [f"index_idx{i} * out_stride{i}" for i in range(rank)]
    )
    code += f"    out_offset = {comp_out}\n"
    code += "    tl.store(out + out_offset, value=cur_inp, mask=mask)\n\n"
    return code


def _generate_wrapper(rank, wrapper_name, kernel_name, code):
    code += f"def {wrapper_name}(inp, dim, index, out, dim_stride, N):\n"
    code += "    inp_shape = inp.shape\n"
    code += "    inp_stride = inp.stride()\n"
    code += "    index_shape = index.shape\n"
    code += "    index_stride = index.stride()\n"
    code += "    out_shape = out.shape\n"
    code += "    out_stride = out.stride()\n"
    code += (
        "    grid = lambda meta: (triton.cdiv(N, meta['BLOCK_SIZE_N']), )\n"
    )
    code += f"    {kernel_name}[grid](\n"
    code += "        inp, index, out,\n"
    for i in range(rank):
        code += f"        inp_shape[{i}],\n"
    for i in range(rank):
        code += f"        index_shape[{i}],\n"
    for i in range(rank):
        code += f"        out_shape[{i}],\n"
    for i in range(rank):
        code += f"        inp_stride[{i}],\n"
    for i in range(rank):
        code += f"        index_stride[{i}],\n"
    for i in range(rank):
        code += f"        out_stride[{i}],\n"
    code += "        dim, dim_stride, N,\n"
    code += "        BLOCK_SIZE_N=512,\n"
    code += "    )\n"
    code += "    return out\n\n"
    return code


def _generate_code(inputs, wrapper_name, kernel_name):
    rank = inputs[0].ndim
    code = ""
    code = _generate_imports(code)
    code = _generate_gather_kernel(rank, kernel_name, code)
    code = _generate_wrapper(rank, wrapper_name, kernel_name, code)
    return code


class GatherFunction:
    def __init__(self):
        self.pid = os.getpid()
        self.overloads = {}

    def __call__(self, *args, **kwargs):
        key = self._arg_key(*args)
        if key in self.overloads:
            overload = self.overloads[key]
        else:
            code = _generate_code(
                args, "_gather_wrapper", "_gather_flaggems_jit_function"
            )
            file_name = f"gather_rank_{key}.py"
            file_path = str(_code_cache_dir() / file_name)
            write_atomic(file_path, code)
            spec = importlib.util.spec_from_file_location(
                f"_gen_module_rank_{key}", file_path
            )
            m = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(m)
            overload = getattr(m, "_gather_wrapper")
            self.overloads[key] = overload
        return overload(*args, **kwargs)

    def _arg_key(self, *args):
        return args[0].ndim


_gather_func = GatherFunction()


def gather(inp, dim, index, out=None, sparse_grad=False):
    if inp.ndim != index.ndim:
        raise IndexError(
            f"self and index must have the same number of dimensions, "
            f"got self.ndim = {inp.ndim} and index.ndim = {index.ndim}"
        )
    if out is None:
        out = torch.empty_like(index, dtype=inp.dtype, device=inp.device)
    dim_stride = inp.stride(dim)
    inp_strided = restride_dim(inp, dim, index.shape)
    N = index.numel()
    _gather_func(inp_strided, dim, index, out, dim_stride, N)
    return out


def gather_backward(grad, self, dim, index, sparse_grad):
    result = grad.new_zeros(self.shape)
    return scatter_(result, dim, index, grad, reduce="add")
```

### Triton IR

`tt.func public @scatter_rank2_add_kernel(...)
...
%1 = tt.addptr %0, %out_offsets_26
  : tensor<128x!tt.ptr<f32>>, tensor<128xi64>

%2 = tt.atomic_rmw fadd, relaxed, gpu, %1, %cur_src_20, %mask_4
  : (tensor<128x!tt.ptr<f32>>, tensor<128xf32>, tensor<128xi1>) -> tensor<128xf32>
### Crash log

```shell
error: unexpected op in ptr sequence
    tl.atomic_add(out + out_offsets, cur_src, mask=mask, sem="relaxed")
    ^

note: see current operation:
tt.atomic_rmw fadd, relaxed, gpu, ... :
(tensor<128x!tt.ptr<f32>>, tensor<128xf32>, tensor<128xi1>) -> tensor<128xf32>

warning: Cannot transform tensor of pointers into a single base pointer with tensor of offsets

error: Dialect `tptr' not found for custom op 'tptr.type_offset'
    %0 = tptr.type_offset i64 : i64

subprocess.CalledProcessError:
buddy-opt ttshared.mlir
  --empty-tensor-to-alloc-tensor
  --one-shot-bufferize=allow-return-allocs-from-loops=true
  --buffer-deallocation-pipeline
  --lower-linalg-to-vir
  --lower-vir-to-vector=vector-width=16
returned non-zero exit status 1.
```

### Additional information

Note: gather_rank2 forward lowers and passes QEMU. This issue is about gather backward, which lowers to scatter reduce="add" and generates tl.atomic_add.

This happens in the TritonShared -> VectorIR lowering stage.

Pipeline:
Triton Python kernel -> TTIR -> TTSharedIR -> VIR -> VectorIR -> LLVM IR -> RISC-V asm -> ELF -> qemu-riscv64

The failure is triggered by atomic dynamic indexed writeback, e.g.:

    tl.atomic_add(out + dynamic_offsets, value, mask=mask)

The generated TTIR contains:

    tt.atomic_rmw fadd ... tensor<...x!tt.ptr<f32>>

Two kernels hit the same issue:

1. TokenEmbedding / embedding_dense_backward
   File:
   rvv_test_qemu/TokenEmbedding/embedding_dense_backward/embedding_dense_backward.py

   Failing source line:
       tl.atomic_add(gw_ptrs, go, mask=mask)

   TTIR dump:
   rvv_test_qemu/TokenEmbedding/embedding_dense_backward/output/dump/tt.mlir

2. Gather backward / scatter reduce="add"
   gather_backward lowers to scatter_(..., reduce="add"), which generates:

       tl.atomic_add(out + inp_offsets, cur_src, mask=mask, sem='relaxed')

   Forward gather with dynamic indexed load/store passes QEMU correctly.
   The failure only appears on the atomic add writeback path.

So the issue does not seem to be ordinary dynamic indexing itself. Plain dynamic load/store can lower and run, but dynamic tensor-pointer atomic add fails during TTSharedIR -> VectorIR because tptr-related IR remains and buddy-opt does not recognize the tptr dialect.

**Comment by maochuanjun** [comment]
Thanks for the detailed report. This has been fixed on current main by PR #44.

I revalidated the original failure pattern end-to-end on RV64/QEMU, including the i64-offset path rather than only the default FlagGems tests:
  - Used the generated rank-2 scatter-add kernel with INT32_OFFSET=False.
  - The TTIR contains tensor<128xi64> dynamic pointer offsets and tt.atomic_rmw fadd.
  - Tested colliding updates: four source rows containing 1, 2, 3, and 4 all scatter to the same output row. QEMU produced 10 for every element of that row and zero elsewhere.

  - The generated ELF exited with status 0 and reported PASS.
  - The lowering contains 4 tt.atomic_rmw fadd operations, 4 memref.atomic_rmw addf operations, and 4 LLVM atomicrmw fadd operations, with no residual tptr.* operations.

  - The ELF is an ELF64 RISC-V executable, and its disassembly contains the expected floating-point add with lr.w.aq / sc.w.rl atomic retry loops.

  I also tested the other reproducer mentioned here, embedding_dense_backward, with i64 indices and colliding gradient updates. It also passed QEMU with the expected result.

  As additional regression coverage, the repository gather/scatter pytest suites pass (25 passed), and both atomic lowering lit tests pass.

  Therefore both the original TTShared/Vector lowering failure and the resulting RISC-V/QEMU execution have been verified as fixed. This issue should be closed as resolved.
